propolis/src/hw/pci/bridge.rs

```diff
@@ -110,41 +100,218 @@ impl Bridge {
+                // (SS3.2.5.1).
+                StdCfgReg::Bar(_) => {}
+
+                // Expansion ROMs are not supported.
+                StdCfgReg::ExpansionRomAddr => {}
+
+                // These registers appear in type-0 PCI headers, but not bridge
+                // headers.
+                StdCfgReg::MaxLatency
+                | StdCfgReg::MinGrant
+                | StdCfgReg::CardbusPtr => {
+                    panic!("Unexpected register type {:?}", id);
                 }
-                _ => self.inner.lock().unwrap().cfg_std_write(id, wo, ctx),
             },
-        })
+
+            // Writable bridge registers.
+            BridgeReg::PrimaryBus => {
+                let mut guard = self.inner.lock().unwrap();
+                guard.primary_bus = BusNum::new(wo.read_u8()).unwrap()
+            }
+            BridgeReg::SecondaryBus => {
+                let mut guard = self.inner.lock().unwrap();
+                guard.set_secondary_bus(BusNum::new(wo.read_u8()).unwrap())
+            }
+            BridgeReg::SubordinateBus => {
+                let mut guard = self.inner.lock().unwrap();
+                guard.subordinate_bus = BusNum::new(wo.read_u8()).unwrap()
+            }
+            BridgeReg::MemoryBase => {
+                let mut guard = self.inner.lock().unwrap();
+                guard.memory_base = wo.read_u16();
+            }
+            BridgeReg::MemoryLimit => {
+                let mut guard = self.inner.lock().unwrap();
+                guard.memory_limit = wo.read_u16();
+            }
+
+            // Read-only bridge registers.
+            BridgeReg::SecondaryLatencyTimer => {}
+            BridgeReg::IoBase | BridgeReg::IoLimit => {}
+            BridgeReg::SecondaryStatus => {}
+            BridgeReg::PrefetchableMemoryBase
+            | BridgeReg::PrefetchableMemoryLimit => {}
+            BridgeReg::PrefetchableMemoryBaseUpper
+            | BridgeReg::PrefetchableMemoryLimitUpper => {}
+            BridgeReg::IoBaseUpper | BridgeReg::IoLimitUpper => {}
+            BridgeReg::BridgeControl => {}
+        }
+    }
+}
+
+impl Endpoint for Bridge {
+    fn attach(&self, attachment: Attachment) {
+        let mut inner = self.inner.lock().unwrap();
+        let _old = inner.attachment.replace(attachment);
+        assert!(_old.is_none());
+    }
+
+    fn cfg_rw(&self, mut rwo: RWOp, ctx: &DispCtx) {
+        self.cfg_map.process(&mut rwo, |id, rwo| match id {
+            CfgReg::Std => {
+                self.cfg_header_rw(rwo, ctx);
+            }
+            _ => {
+                panic!(
+                    "Unexpected read of bridge config space with ID {:?}",
+                    id
+                )
+            }
+        });
+    }
+
+    fn bar_rw(&self, _bar: BarN, _rwo: RWOp, _ctx: &DispCtx) {
+        // The BARs in the PCI bridge are read-only, so nothing should ever
+        // try to dispatch an I/O to a region defined in a bridge's BAR.
```

Comment:
nit: since the BARs are read-only, they shouldn't actually be defining any regions

Reply:
Updated the comment here--let me know if you had something else in mind.